File: bindings/python/examples/scan_ids_demo.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import time

from motorbridge import Controller, RID_CTRL_MODE


def _parse_id(text: str) -> int:
    return int(text, 0)


def _hx(v: int) -> str:
    return f"0x{v:03X}"


def main() -> None:
    p = argparse.ArgumentParser(
        description="Fast CAN-bus scan demo (register probe + enable fallback)"
    )
    p.add_argument("--channel", default="can0")
    p.add_argument("--model", default="4310")
    p.add_argument("--start-id", default="0x001")
    p.add_argument("--end-id", default="0x7FF")
    p.add_argument("--feedback-base", default="0x10")
    p.add_argument("--timeout-ms", type=int, default=80)
    p.add_argument("--sleep-ms", type=int, default=2)
    p.add_argument("--verbose", type=int, default=0)
    args = p.parse_args()

    start_id = _parse_id(args.start_id)
    end_id = _parse_id(args.end_id)
    feedback_base = _parse_id(args.feedback_base)

    if start_id < 0x001 or end_id > 0x7FF or start_id > end_id:
        raise SystemExit("invalid range: use 0x001..0x7FF and start <= end")

    print(
        f"scan start: channel={args.channel} model={args.model} "
        f"range={_hx(start_id)}..{_hx(end_id)} feedback-base={_hx(feedback_base)} "
        f"timeout-ms={args.timeout_ms}"
    )

    hits = 0
    with Controller(args.channel) as ctrl:
        for mid in range(start_id, end_id + 1):
            fid = feedback_base + (mid & 0x0F)
            ok = False
            try:
                m = ctrl.add_damiao_motor(mid, fid, args.model)
                try:
                    _ = m.get_register_u32(RID_CTRL_MODE, args.timeout_ms)
                    ok = True
                except Exception as e_reg:
                    if args.verbose:
                        print(
                            f"[probe-reg-miss] motor-id={_hx(mid)} feedback-id={_hx(fid)} err={e_reg}"
                        )
                    try:
                        m.enable()
                        m.request_feedback()
                        for _ in range(3):
                            ctrl.poll_feedback_once()
                            if m.get_state() is not None:
                                ok = True
                                break
                            time.sleep(0.005)
                    except Exception as e_en:
                        if args.verbose:
                            print(
                                f"[probe-enable-miss] motor-id={_hx(mid)} feedback-id={_hx(fid)} err={e_en}"
                            )
                finally:
                    m.close()
            except Exception as e_add:
                if args.verbose:
                    print(f"[add-motor-miss] motor-id={_hx(mid)} feedback-id={_hx(fid)} err={e_add}")

            if ok:
                print(f"[hit] motor-id={_hx(mid)} feedback-id={_hx(fid)}", flush=True)
                hits += 1

            if args.sleep_ms > 0:
                time.sleep(args.sleep_ms / 1000.0)

    print(f"scan done, hits={hits}")


if __name__ == "__main__":
    main()

File: bindings/cpp/examples/scan_ids_demo.cpp
#include <chrono>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>
#include <thread>

#include "motorbridge/motorbridge.hpp"

static uint16_t parse_u16(const std::string& s) {
  return static_cast<uint16_t>(std::stoul(s, nullptr, 0));
}

static std::string hex3(uint16_t v) {
  std::ostringstream oss;
  oss << "0x" << std::hex << std::uppercase << std::setw(3) << std::setfill('0') << v;
  return oss.str();
}

int main(int argc, char** argv) {
  std::string channel = "can0";
  std::string model = "4310";
  uint16_t start_id = 0x001;
  uint16_t end_id = 0x7FF;
  uint16_t feedback_base = 0x10;
  uint32_t timeout_ms = 80;
  uint32_t sleep_ms = 2;
  int verbose = 0;

  for (int i = 1; i < argc; ++i) {
    std::string k = argv[i];
    if (k == "--channel" && i + 1 < argc) channel = argv[++i];
    else if (k == "--model" && i + 1 < argc) model = argv[++i];
    else if (k == "--start-id" && i + 1 < argc) start_id = parse_u16(argv[++i]);
    else if (k == "--end-id" && i + 1 < argc) end_id = parse_u16(argv[++i]);
    else if (k == "--feedback-base" && i + 1 < argc) feedback_base = parse_u16(argv[++i]);
    else if (k == "--timeout-ms" && i + 1 < argc) timeout_ms = static_cast<uint32_t>(std::stoul(argv[++i]));
    else if (k == "--sleep-ms" && i + 1 < argc) sleep_ms = static_cast<uint32_t>(std::stoul(argv[++i]));
    else if (k == "--verbose" && i + 1 < argc) verbose = std::stoi(argv[++i]);
    else if (k == "--help") {
      std::cout
          << "Usage: ./scan_ids_demo [--channel can0] [--model 4310] [--start-id 0x001] [--end-id 0x7FF]\n"
          << "                      [--feedback-base 0x10] [--timeout-ms 80] [--sleep-ms 2] [--verbose 0|1]\n";
      return 0;
    }
  }

  if (start_id < 0x001 || end_id > 0x7FF || start_id > end_id) {
    std::cerr << "invalid range: use 0x001..0x7FF and start <= end\n";
    return 2;
  }

  try {
    motorbridge::Controller ctrl(channel);
    int hits = 0;

    std::cout << "scan start: channel=" << channel << " model=" << model << " range=" << hex3(start_id)
              << ".." << hex3(end_id) << " feedback-base=" << hex3(feedback_base)
              << " timeout-ms=" << timeout_ms << "\n";

    for (uint16_t mid = start_id; mid <= end_id; ++mid) {
      uint16_t fid = static_cast<uint16_t>(feedback_base + (mid & 0x0F));
      bool ok = false;

      try {
        auto m = ctrl.add_damiao_motor(mid, fid, model);
        try {
          (void)m.get_register_u32(motorbridge::RID_CTRL_MODE, timeout_ms);
          ok = true;
        } catch (const std::exception& e) {
          if (verbose) {
            std::cerr << "[probe-reg-miss] motor-id=" << hex3(mid) << " feedback-id=" << hex3(fid)
                      << " err=" << e.what() << "\n";
          }
        }
        if (!ok) {
          try {
            m.enable();
            m.request_feedback();
            for (int t = 0; t < 3; ++t) {
              ctrl.poll_feedback_once();
              if (m.get_state().has_value()) {
                ok = true;
                break;
              }
              std::this_thread::sleep_for(std::chrono::milliseconds(5));
            }
          } catch (const std::exception& e) {
            if (verbose) {
              std::cerr << "[probe-enable-miss] motor-id=" << hex3(mid) << " feedback-id=" << hex3(fid)
                        << " err=" << e.what() << "\n";
            }
          }
        }
      } catch (const std::exception& e) {
        if (verbose) {
          std::cerr << "[add-motor-miss] motor-id=" << hex3(mid) << " feedback-id=" << hex3(fid)
                    << " err=" << e.what() << "\n";
        }
      }

      if (ok) {
        std::cout << "[hit] motor-id=" << hex3(mid) << " feedback-id=" << hex3(fid) << "\n";
        std::cout.flush();
        ++hits;
      }

      if (sleep_ms > 0) std::this_thread::sleep_for(std::chrono::milliseconds(sleep_ms));
      if (mid == 0x7FF) break;
    }

    std::cout << "scan done, hits=" << hits << "\n";
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "fatal: " << e.what() << "\n";
    return 1;
  }
}

